Handle GenerateName in util.CreateOrUpdate

tpantelis opened the pull request:
If `Name` is empty and `GenerateName` is set in the target resource, use the `List` method to obtain the existing resource by label selector. The motivating use case is for transforming `EndpointSlices` in **lighthouse** via a resource syncer. Instead of constructing a well-known name by concatenating various fields and risking exceeding the max length, we can use the `GenerateName` functionality. This means we need to find the target by some other means than its name. Using List by its labels makes sense although it assumes the labels uniquely identify the target resource. In order to use the `GenerateName` functionality, the caller has to ensure label uniqueness.

This requires adding a `List` method to the `resource.Interface` ([first commit](https://github.com/submariner-io/admiral/pull/687/commits/ab6549ec997088f21c7db5d8129fc226e68374a4)). In addition, add support for `GenerateName` in the fake dynamic client.

tpantelis committed:
Add List method to resource.Interface

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

skitt commented on pkg/federate/base_federator.go:
```diff
@@ -65,9 +72,26 @@ func (f *baseFederator) Delete(obj runtime.Object) error {
 
+	ctx := context.TODO()
+
+	if toDelete.GetName() == "" && toDelete.GetGenerateName() != "" {
```
The Kubernetes APIs don’t support `GenerateName` on delete operations, do we really need this?

tpantelis replied:
This API isn't trying to mimic/mirror the K8s delete API. The scenario here is the caller wants to delete a resource where a well-known name isn't known, ie it was generated via `GenerateName`.  The motivating use case is for transforming K8s `EndpontSlices`  to LH `EndpontSlices` via a resource syncer. Instead of constructing a well-known name by concatenating various fields and risking exceeding the max length,  we use the `GenerateName` functionality. However, on update and delete, this means we need to find the target resource via `List` by its unique labels derived from the source resource. We could do this in the caller but I thought to generalize it here and in `CreateOrUpdate` and codify the intention implicitly in the passed resource. We could codify it in the function signatures instead to make it  explicit. The other option is to have the caller obtain the target resource and not change anything in admiral.

skitt replied:
Ah right, that makes sense. My main concern then with the generalisation is that it should apply generally — it might be sufficient to return an error if the `List` calls return more than one result...

tpantelis replied:
@skitt and I discussed offline - we don't think handling `GenerateName` semantics fits here so we'll let the client handle the lookup and set the name.

tpantelis replied:
This code was removed.

skitt commented on pkg/util/create_or_update_test.go:
```diff
@@ -156,6 +160,19 @@ var _ = Describe("CreateOrUpdate function", func() {
+		Context("and GenerateName is set", func() {
+			BeforeEach(func() {
+				t.pod.Name = ""
+				t.pod.GenerateName = "name-prefix"
```
Typically,

```suggestion
				t.pod.GenerateName = "name-prefix-"
```

tpantelis replied:
yeah

skitt commented on pkg/util/create_or_update_test.go:
```diff
@@ -156,6 +160,19 @@ var _ = Describe("CreateOrUpdate function", func() {
+				Expect(createOrUpdate(util.OperationResultCreated)).To(Succeed())
+				t.verifyPod()
+			})
+		})
```
It would be worth checking that a name was generated and starts with the given prefix.

tpantelis replied:
Done

skitt commented on pkg/util/create_or_update_test.go:
```diff
@@ -338,6 +360,23 @@ func (t *createOrUpdateTestDriver) testUpdate(doUpdate func(util.OperationResult
 			tests.EnsureNoActionsForResource(t.testingFake, "pods/status", "update")
 		})
 
+		Context("and GenerateName is set", func() {
```
I don’t think the Kubernetes APIs support `GenerateName` on update operations.

tpantelis replied:
See above comment re: delete.

tpantelis committed:
Handle GenerateName in util.CreateOrUpdate

If Name is empty and GenerateName is set in the target resource, use
the List method to obtain the existing resource by label selector.

The motivating use case is for transforming EndpontSlices in lighthouse
via a resource syncer. Instead of constructing a well-known name by
concatenating various fields and risking exceeding the max length, we can
use the GenerateName functionality. This means we need to find the target
by some other means than its name. Using List by its labels makes sense
although it assumes the labels uniquely identify the target resource.
In order to use the GenerateName functionality, the caller has to ensure
label uniqueness.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>
